packages/algorand-js/src/interpreter/opcode-list.ts

```diff
@@ -545,13 +581,57 @@ export class Substring3 extends Op {
     const start = this.assertBigInt(stack.pop());
     const end = this.assertBigInt(stack.pop());
 
-    if (end < start) {
-      throw new TealError(ERRORS.TEAL.SUBSTRING_END_BEFORE_START);
-    }
-    if (start > byteString.length || end > byteString.length) {
-      throw new TealError(ERRORS.TEAL.SUBSTRING_RANGE_BEYOND);
+    const subString = this.subString(start, end, byteString);
+    stack.push(subString);
+  }
+}
+
+/** Pseudo-Ops **/
+
+// push integer to stack
+export class Int extends Op {
+  readonly uint64: bigint;
+
+  constructor (uint64: bigint) {
+    super();
+    this.uint64 = uint64;
+  }
+
+  execute (stack: TEALStack): void {
+    stack.push(this.uint64);
+  }
+}
+
+// push bytes to stack
+export class Byte extends Op {
+  readonly str: string;
+  readonly encoding?: EncodingType;
+
+  constructor (str: string, encoding?: EncodingType) {
+    super();
+    this.str = str;
+    if (encoding !== undefined) {
+      this.encoding = encoding;
     }
+  }
+
+  execute (stack: TEALStack): void {
+    const buffer = convertToBuffer(this.str, this.encoding);
+    stack.push(new Uint8Array(buffer));
+  }
+}
 
-    stack.push(byteString.slice(Number(start), Number(end)));
+// decodes algorand address to bytes and pushes to stack
+export class Addr extends Op {
+  readonly addr: string;
+
+  constructor (addr: string) {
+    super();
+    this.addr = addr; // parser should verify the addr first
+  }
+
+  execute (stack: TEALStack): void {
+    const addr = decodeAddress(this.addr);
```

Comment:
does `decodeAddress` verifies address too ?

Reply:
No. I think we discussed parser will verify the address before execution (for better speed). Did you mean something else?

Reply:
No, i thought we were doing all operations related to pseudo-opcodes in pseudo-opcodes

Reply:
let's leave, will cover in parser